core/blockchain_arbitrum.go

```diff
@@ -19,14 +19,39 @@ package core
 
 import (
 	"fmt"
+	"sync"
 	"time"
 
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
 	"github.com/ethereum/go-ethereum/core/state"
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/log"
 	"github.com/ethereum/go-ethereum/rpc"
 )
 
+func (bc *BlockChain) FlushTrieDB(advanceBlockChainMutex *sync.Mutex, capLimit common.StorageSize) error {
+	advanceBlockChainMutex.Lock()
+	defer advanceBlockChainMutex.Unlock()
+
+	if bc.triedb.Scheme() == rawdb.HashScheme {
+		err := bc.triedb.Cap(capLimit)
+		if err != nil {
+			return err
+		}
+	}
+
+	err := bc.triedb.Commit(bc.CurrentBlock().Root, true)
```

Comment:
nitpick: I wonder if we couldn't first commit current block root and then cap remaining dirties cache - that way the capLimit parameter will be more precise.
 
 Currently, after `FlushTireDB` triedb size is:
 `sizeAfter = min(sizeBefore, capLimit) - dirtyNodesFromCurrentRootNotFlushedWithCap` 
  
  If we committed first then we'd have:
  `sizeAfter = min(sizeBefore - dirtyNodesFromCurrentRoot, capLimit)`

Reply:
Done.